File: tests/fixtures/vulnerable.go
package main

import (
	"crypto/ecdh"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/md5"
	"crypto/rand"
	"crypto/rsa"
	"crypto/tls"
	"fmt"
	mathrand "math/rand"
	"net/http"
	"os/exec"
)

func vulnerable() {
	userInput := getUserInput()

	// 1. go/no-sql-injection — string concat (Critical)
	query1 := "SELECT * FROM users WHERE id = " + userInput

	// 2. go/no-sql-injection — fmt.Sprintf (Critical)
	query2 := fmt.Sprintf("SELECT * FROM users WHERE id = %s", userInput)

	// 3. go/no-command-injection (Critical)
	exec.Command(userInput)

	// 4. go/no-hardcoded-secret (High)
	apiKey := "sk-live-abcdef123456789"

	// 5. go/no-weak-crypto (Medium) — import already triggers, plus usage:
	md5.New()

	// 6. go/no-ssrf (High)
	http.Get(userInput)

	// 7. go/no-ssrf (High) via NewRequest
	http.NewRequest("GET", userInput, nil)

	// 8. go/net-http-no-timeout (Medium)
	http.ListenAndServe(":8080", nil)

	// 9. go/insecure-tls-skip-verify (High)
	transport := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}

	// 9b. go/missing-ssl-minversion (Medium) — separate config so it is not
	// masked by the InsecureSkipVerify same-literal skip.
	tlsConfig := &tls.Config{ServerName: "example.com"}

	// 10. go/cookie-missing-secure (Medium)
	http.SetCookie(w, &http.Cookie{Name: "sid", Value: userInput, HttpOnly: true})

	// 11. go/cookie-missing-httponly (Medium)
	http.SetCookie(w, &http.Cookie{Name: "prefs", Value: userInput, Secure: true})

	// 12. go/math-random-used (Medium)
	insecureCode := mathrand.Intn(1000000)

	// 13. go/no-unsafe-deserialization — gob.NewDecoder (High)
	dec := gob.NewDecoder(conn)

	// 14. go/no-unsafe-deserialization — yaml.Unmarshal into interface{} (High)
	var out interface{}
	yaml.Unmarshal(data, new(interface{}))

	// 15. go/jwt-no-verify — jwt.ParseUnverified (Critical)
	jwt.ParseUnverified(tokenStr, &jwt.StandardClaims{})

	// 16. go/jwt-no-verify — jwt.Parse with nil key function (Critical)
	jwt.Parse(tokenStr, nil)

	// 17. go/jwt-hardcoded-secret (High)
	jwt.Parse(tokenStr, func(t *jwt.Token) (interface{}, error) { return []byte("my-secret-key"), nil })

	_ = query1
	_ = query2
	_ = apiKey
	_ = transport
	_ = tlsConfig
	_ = insecureCode
	_ = dec
}

func getUserInput() string {
	return "malicious"
}

// go/pq-vulnerable-crypto
func pqVulnerable() {
	key, _ := rsa.GenerateKey(rand.Reader, 2048)
	_ = key
	ecKey, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	_ = ecKey
	_, _, _ = ed25519.GenerateKey(rand.Reader)
	p256key, _ := ecdh.P256().GenerateKey(rand.Reader)
	_ = p256key
}

File: tests/fixtures/vulnerable.rb
# Ruby test fixture — intentionally vulnerable code for foxguard detection tests

# rb/no-eval
eval(user_input)
instance_eval(params[:code])

# rb/no-command-injection
system(user_input)
exec(user_input)
`#{user_input}`

# rb/no-sql-injection
User.where("name = '#{params[:name]}'")
ActiveRecord::Base.connection.execute("DELETE FROM users WHERE id = #{params[:id]}")

# rb/no-mass-assignment
params.permit!

# rb/no-unsafe-deserialization
Marshal.load(data)
YAML.load(data)

# rb/no-open-redirect
redirect_to params[:url]

# rb/no-csrf-skip
skip_before_action :verify_authenticity_token

# rb/no-html-safe
params[:name].html_safe
raw(user_input)

# rb/no-hardcoded-secret
secret_key = "super-secret-key-12345"
api_token = "sk-live-abcdef123456789"

# rb/no-weak-crypto
Digest::MD5.hexdigest("data")
Digest::SHA1.hexdigest("data")

# rb/no-ssrf
URI.open(user_input)
Net::HTTP.get(user_url)
HTTParty.get(url)
Faraday.get(url)
RestClient.get(url)
open url

# rb/no-path-traversal
File.read(user_input)
File.open(user_input)
IO.read(user_input)
File.write(path, data)
FileUtils.cp(src, dst)
send_file path
